### DISCUSSION TITLE
Better duration support

### DISCUSSION THREAD

**Discussion by NogrTL** [js-fns/date-fns#3666] [Ideas] [answered]
Hi, there.

Working with this library I notice that working with time intervals aka Durations is not well implemented here. You're only options are basically - `intervalToDuration`, `formatDuration` and `formatISODuration` with a lot of limitations and caveats. 

My proposal is to make proper duration API with improvements to exiting. Look around in other libraries, APIs or other languages and make it work here.

Few issues mentioning this: #3560 #3569 #3092 #2788
 
Things I notice so far:
- no option to create new Duration
- getting Duration from `intervalToDuration` can't be used if in case you don't have dates to begin with.
- manually creating Duration object can yield unexpected results. As an example - it doesn't upcast/normalize, so you can end with objects like `{ seconds: 123124512 }` which is technically still a duration but so do the UNIX timestamp and pretty much all the dates🤷‍♂️
- limited formatting options for `formatDuration`
- `Distance` API is an odd finding. Feels like it's relative duration, so I'm not sure why it's a separate thing at all.

**Comment by kossnocorp** [comment, marked as answer]
Hey, I appreciate you raising the issue, but unfortunately, it's not very actionable. I understand it doesn't do enough for you, but it will only be possible to match your expectations if you're clear about them.

I'm not going to promise to move the needle personally, but some folks would love to work on new features, and if you can take time and describe the functionality you wish to have, what features are missing, and which bugs you want to get fixed, I'll be happy to discuss and plan those.

Abstract concerns without full context and examples, and suggestions to look around and improve the library won't cut it.

**Comment by NogrTL** [comment]
Something like API proposal will do?

I'm actually willing to help implement it. Just didn't want to start work without notice.

**Reply by kossnocorp** [reply]
@ckcherry23 started working on the proposals, so here's what we have right now:

- `durationTo`: https://github.com/date-fns/date-fns/issues/3691
- `normalizeDuration`: https://github.com/date-fns/date-fns/issues/3692
- Format duration options: https://github.com/orgs/date-fns/discussions/3693
- Compare duration functions: https://github.com/date-fns/date-fns/issues/3694
- Summing duration functions: https://github.com/date-fns/date-fns/issues/3695
- `parseISODuration`: https://github.com/date-fns/date-fns/pull/3151

Feel free to chime in and also share your additional ideas and suggestions here.

**Comment by nolde** [comment]
Given that [Temporal.Duration](https://tc39.es/proposal-temporal/docs/duration.html) is currently in stage 3, it might be a good idea to give it some extra attention.
